modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+
+			Self::deposit_event(Event::CallDispatched {
+				task_result: result.map(|_| ()).map_err(|e| e.error),
+			});
+			Ok(())
+		}
+
+		/// Remove Query that has expired so chain state does not bloat. This rewards the oracle
+		/// with half the query fee
+		///
+		/// - `query_index`: Unique index that is mapped to a particular query
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn remove_expired_call(origin: OriginFor<T>, query_index: QueryIndex) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let verifiable_call = ActiveQuery::<T>::take(query_index).ok_or(Error::<T>::NoMatchingCall)?;
+			// Make sure query is expired
+			ensure!(
+				verifiable_call.expiry <= T::BlockNumberProvider::current_block_number(),
+				Error::<T>::QueryNotExpired
+			);
+
+			// Gives half of reward for clearing expired query from storage
+			let reward: Balance = verifiable_call.oracle_reward.div(2u32);
+			T::Currency::transfer(&Self::account_id(), &who, reward, ExistenceRequirement::KeepAlive)?;
+
+			Self::deposit_event(Event::<T>::StaleQueryRemoved { query_index });
+			Ok(())
+		}
+
+		/// Cancel query before it has been serviced, only the account that created the query can
+		/// cancel it.
+		///
+		/// - `query_index`: Unique index mapped to a particular query
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn cancel_query(origin: OriginFor<T>, query_index: QueryIndex) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::cancel_task(&who, query_index)?;
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	// Returns pallet account
+	pub fn account_id() -> T::AccountId {
+		T::PalletId::get().into_account()
+	}
+}
+
+impl<T: Config> ForeignChainStateQuery<T::AccountId, T::VerifiableTask> for Pallet<T> {
+	#[transactional]
+	fn query_task(who: T::AccountId, length_bound: usize, dispatchable_call: T::VerifiableTask) -> DispatchResult {
+		let call_len = dispatchable_call.using_encoded(|x| x.len());
+		ensure!(call_len <= length_bound, Error::<T>::TooLargeVerifiableCall);
+		T::Currency::transfer(
+			&who,
+			&Self::account_id(),
+			T::QueryFee::get(),
+			ExistenceRequirement::KeepAlive,
+		)?;
+
+		let expiry = T::BlockNumberProvider::current_block_number().saturating_add(T::QueryDuration::get());
+		let verifiable_call = VerifiableCall {
+			dispatchable_call,
+			expiry,
+			oracle_reward: T::QueryFee::get(),
+			account_id: who,
+		};
+
+		let index = QueryCounter::<T>::get();
+		// Increment counter by one
+		QueryCounter::<T>::put(index.checked_add(One::one()).ok_or(ArithmeticError::Overflow)?);
+
+		ActiveQuery::<T>::insert(index, verifiable_call);
+		Self::deposit_event(Event::CreateActiveQuery { index, expiry });
+		Ok(())
+	}
+
+	#[transactional]
+	fn cancel_task(who: &T::AccountId, index: QueryIndex) -> DispatchResult {
+		let task = ActiveQuery::<T>::take(index).ok_or(Error::<T>::NoMatchingCall)?;
+		ensure!(who == &task.account_id, Error::<T>::NotQueryAccount);
```

Comment:
this isn't important. we don't need permission check on this level. we will assume the caller will do it